lib/puppet/util/posix.rb

```diff
@@ -1,5 +1,48 @@
 # Utility methods for interacting with POSIX objects; mostly user and group
 module Puppet::Util::POSIX
+  require 'ffi'
+  module GrpLibC
+    extend FFI::Library
+    ffi_lib FFI::Library::LIBC
+    begin
+      attach_function :getgrouplist, [:string, :int, :pointer, :pointer], :int
+    rescue FFI::NotFoundError => e
+      puts e.message
+    end
+
+    def self.user_groups(user)
+      gid = Puppet::Etc.getpwnam(user).gid
+      ngroups_ptr = FFI::MemoryPointer.new(:int)
+      ngroups = 16
+      ngroups_ptr.write_int(ngroups)
+      groups_ptr = FFI::MemoryPointer.new(:uint, ngroups)
+
+      # getgrouplist updates ngroups_ptr to num required.
+      ret = GrpLibC.getgrouplist(user, gid, groups_ptr, ngroups_ptr)
+
+      # # FIXME: some systems (like Darwin) have a bug where they
+      # # never increase ngroups_ptr
+      # while ret < 0
+      #   if (ngroups == ngroups_ptr.get_int(0))
+      #     ngroups *= 2;
+      #     ngroups_ptr.write_int(ngroups)
+      #   end
+      #   groups_ptr.free if groups_ptr
+      #   groups_ptr = FFI::MemoryPointer.new(:uint, ngroups_ptr.get_int(0))
+      #   ret = GrpLibC.getgrouplist(user, gid, groups_ptr, ngroups_ptr)
+      # end
+      if ret < 0
+        groups_ptr.free if groups_ptr
+        groups_ptr = FFI::MemoryPointer.new(:uint, ngroups_ptr.get_int(0))
```

Comment:
MemoryPointer inherits from AbstractMemory, so you can do things like `ngroups_ptr.read_uint` and drop the '0' offset.